File: Assets/Plugins/Tim/IOS/TIMJsonUtil.m
//
//  TIMJsonUtil.m
//  Unity-iPhone
//
//  Created by 李政奇 on 2019/11/7.
//

#import "TIMJsonUtil.h"
#import "YYModel.h"


@implementation TIMJsonUtil


+ (NSString*) ErrorToJson:(TAction)action code:(int)code desc:(NSString*) desc{
    NSMutableDictionary *dict = [NSMutableDictionary dictionary];
    [dict setObject:@(TStateFailed) forKey:@"state"];
    [dict setObject:@(action) forKey:@"action"];
    [dict setObject:@(code) forKey:@"code"];
    [dict setObject:desc forKey:@"desc"];
    
    return [dict yy_modelToJSONString];
}
+ (NSString*) SuccessToJson:(TAction) action desc:(NSString*)desc{
    NSMutableDictionary *dict = [NSMutableDictionary dictionary];
    [dict setObject:@(TStateSuccess) forKey:@"state"];
    [dict setObject:@(action) forKey:@"action"];
    [dict setObject:desc forKey:@"desc"];
    
    return [dict yy_modelToJSONString];
}
+ (NSString*) NotifyToJson:(TAction)action desc:(NSString*)desc{
    NSMutableDictionary *dict = [NSMutableDictionary dictionary];
    [dict setObject:@(TStateNotify) forKey:@"state"];
    [dict setObject:@(action) forKey:@"action"];
    [dict setObject:desc forKey:@"desc"];

    return [dict yy_modelToJSONString];
}
//用户信息
+ (NSString*) ProfileToJson:(TAction)action profile:(TIMUserProfile*)profile{
    NSMutableDictionary *dict = [NSMutableDictionary dictionary];
        [dict setObject:@(TStateSuccess) forKey:@"state"];
        [dict setObject:@(action) forKey:@"action"];
    
        NSMutableDictionary * obj= [NSMutableDictionary dictionary];
        [obj setObject:profile.identifier forKey:@"identifier"];
        [obj setObject:profile.nickname forKey:@"nickName"];
        [obj setObject:profile.faceURL forKey:@"faceUrl"];
        [obj setObject:@(profile.gender) forKey:@"gender"];
        [obj setObject:@(profile.birthday) forKey:@"birthday"];
        [obj setObject:@(profile.allowType) forKey:@"allow"];
        [obj setObject:profile.selfSignature forKey:@"selfSignature"];
    
        [dict setObject:obj forKey:@"json"];

        return [dict yy_modelToJSONString];
}

//好友处理结果
+ (NSString*) FriendResultToJson:(TAction) action timFriend:(TIMFriendResult*)timFriend{
    NSMutableDictionary *dict = [NSMutableDictionary dictionary];
    [dict setObject:@(TStateSuccess) forKey:@"state"];
    [dict setObject:@(action) forKey:@"action"];
    
    NSMutableDictionary * obj= [NSMutableDictionary dictionary];
    [obj setObject:timFriend.identifier forKey:@"identifier"];
    [obj setObject:@(timFriend.result_code) forKey:@"resultCode"];
    [obj setObject:timFriend.result_info forKey:@"resultInfo"];
    
    [dict setObject:obj forKey:@"json"];
    return [dict yy_modelToJSONString];
}
  //好友列表
+ (NSString*) FriendListToJson:(NSArray<TIMFriend*> *) timFriends{
    NSMutableDictionary *dict = [NSMutableDictionary dictionary];
    [dict setObject:@(TStateSuccess) forKey:@"state"];
    [dict setObject:@(TActionFriendsList) forKey:@"action"];
    
    NSMutableArray *array=[NSMutableArray array];
    for (TIMFriend* result in timFriends) {
        NSMutableDictionary * obj= [NSMutableDictionary dictionary];
        [obj setObject:result.profile.identifier forKey:@"identifier"];
        [obj setObject:result.profile.nickname forKey:@"nickName"];
        [obj setObject:result.remark forKey:@"remark"];
        [obj setObject:result.profile.faceURL forKey:@"faceUrl"];
        [obj setObject:@(result.profile.gender) forKey:@"gender"];
        [obj setObject:@(result.profile.birthday) forKey:@"birthday"];
        [obj setObject:@(result.profile.allowType) forKey:@"allow"];
        [obj setObject:result.profile.selfSignature forKey:@"selfSignature"];
        
        [array addObject:obj];
    }
    
    [dict setObject:array forKey:@"json"];
    return [dict yy_modelToJSONString];
}
//未决列表
+ (NSString*) PendentListToJson:(TIMFriendPendencyResponse*) response{
    NSMutableDictionary *dict = [NSMutableDictionary dictionary];
    [dict setObject:@(TStateSuccess) forKey:@"state"];
    [dict setObject:@(TActionPendencyList) forKey:@"action"];
    
    NSMutableArray *array=[NSMutableArray array];
    for (TIMFriendPendencyItem* result in response.pendencies) {
        NSMutableDictionary * obj= [NSMutableDictionary dictionary];
        [obj setObject:result.identifier forKey:@"identifier"];
        [obj setObject:result.nickname forKey:@"nickName"];
        [obj setObject:result.addWording forKey:@"addWording"];
        [obj setObject:result.addSource forKey:@"addSource"];
        [obj setObject:@(result.type) forKey:@"type"];
        
        [array addObject:obj];
    }
    
    [dict setObject:array forKey:@"json"];
    return [dict yy_modelToJSONString];
}
//会话信息
+ (NSString*) ConversationToJson:(NSArray<TIMConversation*>*)conversations{
    NSMutableDictionary *dict = [NSMutableDictionary dictionary];
    [dict setObject:@(TStateSuccess) forKey:@"state"];
    [dict setObject:@(TActionConversationRefresh) forKey:@"action"];
    
    NSMutableArray *array=[NSMutableArray array];
    for (TIMConversation* result in conversations) {
        NSMutableDictionary * obj= [NSMutableDictionary dictionary];
        [obj setObject:[result getReceiver] forKey:@"peer"];
        [obj setObject:@([result getType]) forKey:@"type"];
        [obj setObject:[TIMJsonUtil TimMessage2Dic:[result getLastMsg]] forKey:@"lastMsg"];
        
        [array addObject:obj];
    }
    
    [dict setObject:array forKey:@"json"];
    return [dict yy_modelToJSONString];
}
//聊天消息
+ (NSString*) MessageToJson:(TAction)action message:(TIMMessage*)message{
    NSMutableDictionary *dict = [NSMutableDictionary dictionary];
    [dict setObject:@(TStateSuccess) forKey:@"state"];
    [dict setObject:@(action) forKey:@"action"];
    
    [dict setObject:[TIMJsonUtil TimMessage2Dic:message] forKey:@"json"];
    return [dict yy_modelToJSONString];
}
//聊天消息列表
+ (NSString*) MessageListToJson:(TAction) action timMessges:(NSArray<TIMMessage*>*) timMessages{
    NSMutableDictionary *dict = [NSMutableDictionary dictionary];
    [dict setObject:@(TStateSuccess) forKey:@"state"];
    [dict setObject:@(action) forKey:@"action"];
    
    NSMutableArray *array=[NSMutableArray array];
    for (TIMMessage* msg in timMessages) {
        if ([msg isKindOfClass:[TIMMessage class]]) {
            [array addObject:[TIMJsonUtil TimMessage2Dic:msg]];
        }
    }
    
    [dict setObject:array forKey:@"json"];
    return [dict yy_modelToJSONString];
}

//Tim的消息转Json字典
+ (NSDictionary*) TimMessage2Dic:(TIMMessage*) message{
    NSMutableDictionary* dict=[NSMutableDictionary dictionary];
    TIMElem* elem=[message getElem:0];
    
    [dict setObject:message.sender forKey:@"sender"];
    [dict setObject:@(message.isSelf) forKey:@"isSelf"];
    [dict setObject:@([[message getConversation] getType]) forKey:@"conversationType"];
    
    if ([elem isKindOfClass:[TIMTextElem class]]) {
        [dict setObject:@"Text" forKey:@"type"];
        [dict setObject:((TIMTextElem*)elem).text forKey:@"text"];
    }
    else if ([elem isKindOfClass:[TIMImageElem class]]){
        [dict setObject:@"Image" forKey:@"type"];
        [dict setObject:@"" forKey:@"image"];
    }
    else if ([elem isKindOfClass:[TIMVideoElem class]]){
        [dict setObject:@"Video" forKey:@"type"];
        [dict setObject:@"" forKey:@"video"];
    }
    return dict;
}

//Group Event消息
+ (NSString*) GroupEventToJson:(TIMGroupTipsElem*)elem{
    NSMutableDictionary *dict = [NSMutableDictionary dictionary];
     [dict setObject:@(TStateNotify) forKey:@"state"];
     [dict setObject:@(TActionGroupEvent) forKey:@"action"];
    
     NSMutableDictionary *obj = [NSMutableDictionary dictionary];
     [obj setObject:elem.group forKey:@"groupId"];
     [obj setObject:elem.groupName forKey:@"groupName"];
     [obj setObject:@(elem.memberNum) forKey:@"memberNum"];
     [obj setObject:@(elem.type) forKey:@"tipsType"];
     
     [dict setObject:obj forKey:@"json"];
     return [dict yy_modelToJSONString];
}

//创建群
+ (NSString*) GroupToJson:(NSString*) group{
    NSMutableDictionary *dict = [NSMutableDictionary dictionary];
    [dict setObject:@(TStateNotify) forKey:@"state"];
    [dict setObject:@(TActionCreateGroup) forKey:@"action"];
    
    [dict setObject:group forKey:@"json"];
    return [dict yy_modelToJSONString];
}

//群成员
+ (NSString*) GroupMembersToJson:(NSArray<TIMGroupMemberInfo*>*) infoArray{
    NSMutableDictionary *dict = [NSMutableDictionary dictionary];
     [dict setObject:@(TStateSuccess) forKey:@"state"];
     [dict setObject:@(TActionGetGroupMembers) forKey:@"action"];
            
     NSMutableArray *array=[NSMutableArray array];
     for (TIMGroupMemberInfo* mem in infoArray) {
         [array addObject:mem.member];
     }
            
    [dict setObject:array forKey:@"json"];
     return [dict yy_modelToJSONString];
}
+ (NSString*) GroupListToJson:(NSArray<TIMGroupInfo*>*) timInfos{
    NSMutableDictionary *dict = [NSMutableDictionary dictionary];
     [dict setObject:@(TStateSuccess) forKey:@"state"];
     [dict setObject:@(TActionGroupList) forKey:@"action"];
            
     NSMutableArray *array=[NSMutableArray array];
     for (TIMGroupInfo* mem in timInfos) {
         NSMutableDictionary* obj=[NSMutableDictionary dictionary];
         [obj setObject:mem.group forKey:@"groupId"];
         [obj setObject:mem.groupName forKey:@"groupName"];
         [obj setObject:mem.groupType forKey:@"groupType"];
         
         [array addObject:obj];
     }
    
    [dict setObject:array forKey:@"json"];
     return [dict yy_modelToJSONString];
}


@end

File: Assets/Plugins/Tim/Android/TimJsonUtil.java
package com.jiuwei.mylibrary;

import com.tencent.imsdk.TIMConversation;
import com.tencent.imsdk.TIMElem;
import com.tencent.imsdk.TIMElemType;
import com.tencent.imsdk.TIMGroupMemberInfo;
import com.tencent.imsdk.TIMGroupTipsElem;
import com.tencent.imsdk.TIMImageElem;
import com.tencent.imsdk.TIMMessage;
import com.tencent.imsdk.TIMSoundElem;
import com.tencent.imsdk.TIMTextElem;
import com.tencent.imsdk.TIMUserProfile;
import com.tencent.imsdk.ext.group.TIMGroupBaseInfo;
import com.tencent.imsdk.friendship.TIMFriend;
import com.tencent.imsdk.friendship.TIMFriendPendencyItem;
import com.tencent.imsdk.friendship.TIMFriendPendencyResponse;
import com.tencent.imsdk.friendship.TIMFriendResult;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.List;

enum TState
{
    Failed,
    Success,
    Notify,
}

enum TAction
{
    ForceOffline,
    UserSigExpired,
    Connected,
    Disconnected,
    WifiNeedAuth,
    GroupEvent,
    Refresh,
    ConversationRefresh,
    NewMessage,
    Login,
    Logout,
    SelfProfile,
    ModifyProfile,
    UserProfile,
    FriendsList,
    AddFriend,
    DeleteFriend,
    DoResponse,
    PendencyList,
    SendMessage,
    GetMessage,
    CreateGroup,
    ApplyJoinGroup,
    QuitGroup,
    GetGroupMembers,
    GetGroupList
}

public class TimJsonUtil {

    public static String ErrorToJson(TAction action, int code, String desc) {
        JSONObject map = new JSONObject();

        try {
            map.put("state", TState.Failed);
            map.put("action", action);
            map.put("code", code);
            map.put("desc", desc);
        } catch (JSONException e) {
            e.printStackTrace();
        }

        return map.toString();
    }

    public static String SuccessToJson(TAction action,String desc) {
        JSONObject map = new JSONObject();

        try {
            map.put("state", TState.Success);
            map.put("action", action);
            map.put("desc", desc);

        } catch (JSONException e) {
            e.printStackTrace();
        }

        return map.toString();
    }

    public static String NotifyToJson(TAction action, String desc) {
        JSONObject map = new JSONObject();

        try {
            map.put("state", TState.Notify);
            map.put("action", action);
            map.put("desc", desc);
        } catch (JSONException e) {
            e.printStackTrace();
        }

        return map.toString();
    }

    //用户信息
    public static String ProfileToJson(TAction action,TIMUserProfile self) {
        JSONObject map = new JSONObject();

        try {
            map.put("state", TState.Success);
            map.put("action", action);

            JSONObject obj = new JSONObject();
            obj.put("identifier", self.getIdentifier());
            obj.put("nickName", self.getNickName());
            obj.put("faceUrl", self.getFaceUrl());
            obj.put("gender", self.getGender());
            obj.put("birthday", self.getBirthday());
            obj.put("allow", self.getAllowType());
            obj.put("selfSignature", self.getSelfSignature());

            map.put("json", obj);

        } catch (JSONException e) {
            e.printStackTrace();
        }

        return map.toString();
    }

    //好友处理结果
    public static String FriendResultToJson(TAction action, TIMFriendResult timFriend) {
        JSONObject map = new JSONObject();

        try {
            map.put("state", TState.Success);
            map.put("action", action);

            JSONObject obj = new JSONObject();
            obj.put("identifier", timFriend.getIdentifier());
            obj.put("resultCode", timFriend.getResultCode());
            obj.put("resultInfo", timFriend.getResultInfo());

            map.put("json", obj);

        } catch (JSONException e) {
            e.printStackTrace();
        }

        return map.toString();
    }

    //好友列表
    public static String FriendListToJson(List<TIMFriend> timFriends) {

        JSONObject map = new JSONObject();

        try {
            map.put("state", TState.Success);
            map.put("action", TAction.FriendsList);

            JSONArray array = new JSONArray();
            for (TIMFriend friend : timFriends) {
                JSONObject obj = new JSONObject();
                obj.put("identifier", friend.getTimUserProfile().getIdentifier());
                obj.put("nickName", friend.getTimUserProfile().getNickName());
                obj.put("remark", friend.getRemark());
                obj.put("faceUrl", friend.getTimUserProfile().getFaceUrl());
                obj.put("gender", friend.getTimUserProfile().getGender());
                obj.put("birthday", friend.getTimUserProfile().getBirthday());
                obj.put("allow", friend.getTimUserProfile().getAllowType());
                obj.put("selfSignature", friend.getTimUserProfile().getSelfSignature());

                array.put(obj);
            }
            map.put("json", array);

        } catch (JSONException e) {
            e.printStackTrace();
        }

        return map.toString();
    }

    //未决列表
    public static String PendentListToJson(TIMFriendPendencyResponse response) {
        JSONObject map = new JSONObject();

        try {
            map.put("state", TState.Success);
            map.put("action", TAction.PendencyList);

            JSONArray array = new JSONArray();
            List<TIMFriendPendencyItem> timPendents = response.getItems();
            for (TIMFriendPendencyItem item : timPendents) {
                JSONObject obj = new JSONObject();
                obj.put("identifier", item.getIdentifier());
                obj.put("nickName", item.getNickname());
                obj.put("addWording", item.getAddWording());
                obj.put("addSource", item.getAddSource());
                obj.put("type", item.getType());
                array.put(obj);
            }
            map.put("json", array);

        } catch (JSONException e) {
            e.printStackTrace();
        }

        return map.toString();
    }


    //会话信息
    public static String ConversationToJson(List<TIMConversation> conversations) {
        JSONObject map = new JSONObject();

        try {
            map.put("state", TState.Success);
            map.put("action", TAction.ConversationRefresh);

            JSONArray array = new JSONArray();
            for (TIMConversation con : conversations) {
                JSONObject obj = new JSONObject();
                obj.put("peer", con.getPeer());
                obj.put("type", con.getType());
                obj.put("lastMsg", TIMMessage2Dic(con.getLastMsg()));
                array.put(obj);
            }
            map.put("json", array);
        } catch (JSONException e) {
            e.printStackTrace();
        }

        return map.toString();
    }

    //聊天消息
    public static String MessageToJson(TAction action, TIMMessage message) {
        JSONObject map = new JSONObject();

        try {
            map.put("state", TState.Success);
            map.put("action", action);
            map.put("json", TIMMessage2Dic(message));
        } catch (JSONException e) {
            e.printStackTrace();
        }

        return map.toString();
    }

    //聊天消息列表
    public static String MessageListToJson(TAction action, List<TIMMessage> timMessages) {
        JSONObject map = new JSONObject();

        try {
            map.put("state", TState.Success);
            map.put("action", action);

            JSONArray array = new JSONArray();
            for (TIMMessage message : timMessages) {
                JSONObject msg = TIMMessage2Dic(message);

                array.put(msg);
            }
            map.put("json", array);
        } catch (JSONException e) {
            e.printStackTrace();
        }

        return map.toString();
    }

    //Tim的消息转Json
    static JSONObject TIMMessage2Dic(TIMMessage timMessage) {
        JSONObject map = new JSONObject();

        try {
            TIMElem timElem = timMessage.getElement(0);
            TIMElemType type = timElem.getType();

            map.put("type", type);//消息类型
            map.put("conversationType", timMessage.getConversation().getType());//消息类型
            map.put("sender", timMessage.getSender());//发送方ID
            map.put("isSelf", timMessage.isSelf());//是否是自己发送的

            switch (type) {
                case Text://文本
                    TIMTextElem text = (TIMTextElem) timElem;
                    map.put("text", text.getText());//消息
                    break;
                case Image://图片
                    TIMImageElem image = (TIMImageElem) timElem;
                    map.put("image", "");//消息
                    break;
                case Sound://语音
                    TIMSoundElem sound = (TIMSoundElem) timElem;
                    map.put("sound", "");//消息
                    break;
            }
        } catch (JSONException e) {
            e.printStackTrace();
        }

        return map;
    }

    //群组事件
    public static String GroupEventToJson(TIMGroupTipsElem elem) {
        JSONObject map = new JSONObject();

        try {
            map.put("state", TState.Notify);
            map.put("action", TAction.GroupEvent);

            JSONObject obj = new JSONObject();
            obj.put("groupId", elem.getGroupId());
            obj.put("groupName", elem.getGroupName());
            obj.put("memberNum", elem.getMemberNum());
            obj.put("tipsType", elem.getTipsType());

            map.put("json", obj);
        } catch (JSONException e) {
            e.printStackTrace();
        }

        return map.toString();
    }

    //创建群
    public static String GroupToJson(String groupId)
    {
        JSONObject map = new JSONObject();

        try {
            map.put("state", TState.Success);
            map.put("action", TAction.CreateGroup);
            map.put("json", groupId);
        } catch (JSONException e) {
            e.printStackTrace();
        }

        return map.toString();
    }

    //群成员
    public static String GroupMembersToJson(List<TIMGroupMemberInfo> infoList)
    {
        JSONObject map = new JSONObject();

        try {
            map.put("state", TState.Success);
            map.put("action", TAction.GetGroupMembers);

            JSONArray array = new JSONArray();
            for (TIMGroupMemberInfo info : infoList) {
                array.put(info.getUser());
            }
            map.put("json", array);
        } catch (JSONException e) {
            e.printStackTrace();
        }

        return map.toString();
    }

    //群列表
    public static String GroupListToJson(List<TIMGroupBaseInfo> timGroupInfos)
    {
        JSONObject map = new JSONObject();

        try {
            map.put("state", TState.Success);
            map.put("action", TAction.GetGroupList);

            JSONArray array = new JSONArray();
            for (TIMGroupBaseInfo info : timGroupInfos) {
                JSONObject obj=new JSONObject();
                obj.put("groupId",info.getGroupId());
                obj.put("groupName",info.getGroupName());
                obj.put("groupType",info.getGroupType());

                array.put(obj);
            }
            map.put("json", array);
        } catch (JSONException e) {
            e.printStackTrace();
        }

        return map.toString();
    }
}
